File: 59-spiral-matrix-ii/spiral-matrix-ii.c
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** generateMatrix(int n, int* returnSize, int** returnColumnSizes) {
    int** matrix = (int**)malloc(n * sizeof(int*));
    *returnColumnSizes = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        matrix[i] = (int*)malloc(n * sizeof(int));
        (*returnColumnSizes)[i] = n;
    }
    *returnSize = n;

    int num = 1, top = 0, left = 0, bottom = n - 1, right = n - 1;
    while (num <= n * n) {
        for (int i = left; i <= right; i++)
            matrix[top][i] = num++;
        top++;
        for (int i = top; i <= bottom; i++)
            matrix[i][right] = num++;
        right--;
        for (int i = right; i >= left; i--)
            matrix[bottom][i] = num++;
        bottom--;
        for (int i = bottom; i >= top; i--)
            matrix[i][left] = num++;
        left++;
    }
    return matrix;
}

File: 59-spiral-matrix-ii/spiral-matrix-ii.ts
function generateMatrix(n: number): number[][] {
    const matrix = Array.from({ length: n }, () => Array(n).fill(0));
    let num = 1, top = 0, left = 0, bottom = n - 1, right = n - 1;
    while (num <= n * n) {
        for (let i = left; i <= right; i++) matrix[top][i] = num++;
        top++;
        for (let i = top; i <= bottom; i++) matrix[i][right] = num++;
        right--;
        for (let i = right; i >= left; i--) matrix[bottom][i] = num++;
        bottom--;
        for (let i = bottom; i >= top; i--) matrix[i][left] = num++;
        left++;
    }
    return matrix;
};